File: node/lesson-1/src/engine.js
// Shared query engine used by the web UI and the CLI.
//
// Keeps a cached retriever so repeated questions don't re-read and re-rank from
// scratch, rebuilding when the docs folder changes or the retriever type is
// switched. JS is single-threaded, but async I/O lets concurrent web requests
// interleave at `await` points, so index build/load is serialized through a
// promise-chain mutex (the analogue of engine.py's threading.Lock).

import { SYSTEM_PROMPT, buildUserPrompt } from "./prompts.js";
import { getProvider } from "./providers/index.js";
import { buildRetriever } from "./retriever.js";
import { buildIndex, isStale, loadChunks } from "./store.js";

const cache = { retriever: null, key: null };

// Run async sections one at a time: each waits for the previous to settle, and
// the chain itself never rejects so one failure can't wedge later callers.
let lock = Promise.resolve();
function withLock(fn) {
  const result = lock.then(fn, fn);
  lock = result.then(
    () => {},
    () => {}
  );
  return result;
}

export async function refreshIndex(config) {
  // Force a rebuild of the on-disk index and invalidate the retriever cache.
  return withLock(async () => {
    const { chunks, fileCount } = await buildIndex(config);
    cache.retriever = null;
    cache.key = null;
    return { chunks, fileCount };
  });
}

export async function getRetriever(config) {
  // Return a retriever for the current docs, rebuilding only when needed.
  return withLock(async () => {
    if (isStale(config)) {
      await buildIndex(config);
      cache.retriever = null;
    }
    if (cache.retriever === null || cache.key !== config.retriever) {
      const chunks = loadChunks(config);
      cache.retriever = buildRetriever(chunks, config);
      cache.key = config.retriever;
    }
    return cache.retriever;
  });
}

export function dedupSources(hits) {
  const seen = [];
  for (const h of hits) {
    const tag = `${h.source}:${h.page_number}`;
    if (!seen.includes(tag)) seen.push(tag);
  }
  return seen;
}

export async function answerQuestion(config, question) {
  // Retrieve grounding context, call the provider, and return a result object.
  const retriever = await getRetriever(config);
  const hits = retriever.search(question, config.topK);
  const provider = getProvider(config.provider, config);
  const answer = await provider.chat(SYSTEM_PROMPT, buildUserPrompt(question, hits));
  return {
    answer: answer.trim(),
    sources: dedupSources(hits),
    provider: config.provider,
    retriever: retriever.name || config.retriever,
    num_hits: hits.length,
  };
}

File: dotnet/lesson-1/Engine.cs
// Shared query engine used by the CLI and web UI.
// Mirrors localrag/engine.py: a cached retriever rebuilt when docs change or the
// retriever type is switched; thread-safe for concurrent web requests.

using System.Text.Json.Serialization;
using LocalRag.Providers;

namespace LocalRag;

// snake_case JSON keys to match the Python web API and the HTML's fetch code.
public record AnswerResult(
    [property: JsonPropertyName("answer")] string Answer,
    [property: JsonPropertyName("sources")] List<string> Sources,
    [property: JsonPropertyName("provider")] string Provider,
    [property: JsonPropertyName("retriever")] string Retriever,
    [property: JsonPropertyName("num_hits")] int NumHits);

public static class Engine
{
    private static readonly object Lock = new();
    private static IRetriever? _retriever;
    private static string? _key;

    /// <summary>Force a rebuild of the on-disk index and invalidate the retriever cache.</summary>
    public static (List<Chunk> Chunks, int FileCount) RefreshIndex(Config config)
    {
        lock (Lock)
        {
            var (chunks, nFiles) = Store.BuildIndex(config);
            _retriever = null;
            _key = null;
            return (chunks, nFiles);
        }
    }

    /// <summary>Return a retriever for the current docs, rebuilding only when needed.</summary>
    public static IRetriever GetRetriever(Config config)
    {
        lock (Lock)
        {
            if (Store.IsStale(config))
            {
                Store.BuildIndex(config);
                _retriever = null;
            }
            if (_retriever is null || _key != config.Retriever)
            {
                var chunks = Store.LoadChunks(config);
                _retriever = RetrieverFactory.BuildRetriever(chunks, config);
                _key = config.Retriever;
            }
            return _retriever;
        }
    }

    public static List<string> DedupSources(List<Chunk> hits)
    {
        var seen = new List<string>();
        foreach (var h in hits)
        {
            var tag = $"{h.Source}:{h.PageNumber}";
            if (!seen.Contains(tag))
            {
                seen.Add(tag);
            }
        }
        return seen;
    }

    /// <summary>Retrieve grounding context, call the provider, and return a result.</summary>
    public static AnswerResult AnswerQuestion(Config config, string question)
    {
        var retriever = GetRetriever(config);
        var hits = retriever.Search(question, config.TopK);
        var provider = ProviderFactory.GetProvider(config.Provider, config);
        var answer = provider.Chat(Prompts.SystemPrompt, Prompts.BuildUserPrompt(question, hits));
        return new AnswerResult(
            Answer: answer.Trim(),
            Sources: DedupSources(hits),
            Provider: config.Provider,
            Retriever: retriever.Name,
            NumHits: hits.Count);
    }
}
